src/theory/uf/theory_uf_type_rules.cpp

```diff
@@ -160,6 +162,112 @@ TypeNode HoApplyTypeRule::computeType(NodeManager* nodeManager,
   }
 }
 
+TypeNode LambdaTypeRule::computeType(NodeManager* nodeManager,
+                                     TNode n,
+                                     bool check)
+{
+  if (n[0].getType(check) != nodeManager->boundVarListType())
+  {
+    std::stringstream ss;
+    ss << "expected a bound var list for LAMBDA expression, got `"
+       << n[0].getType().toString() << "'";
+    throw TypeCheckingExceptionPrivate(n, ss.str());
+  }
+  std::vector<TypeNode> argTypes;
+  for (TNode::iterator i = n[0].begin(); i != n[0].end(); ++i)
+  {
+    argTypes.push_back((*i).getType());
+  }
+  TypeNode rangeType = n[1].getType(check);
+  return nodeManager->mkFunctionType(argTypes, rangeType);
+}
+
+bool LambdaTypeRule::computeIsConst(NodeManager* nodeManager, TNode n)
+{
+  Assert(n.getKind() == kind::LAMBDA);
+  // get array representation of this function, if possible
+  Node na = FunctionConst::getArrayRepresentationForLambda(n);
+  if (!na.isNull())
+  {
+    Assert(na.getType().isArray());
+    Trace("lambda-const") << "Array representation for " << n << " is " << na
+                          << " " << na.getType() << std::endl;
+    // must have the standard bound variable list
+    Node bvl =
+        NodeManager::currentNM()->getBoundVarListForFunctionType(n.getType());
+    if (bvl == n[0])
+    {
+      // array must be constant
+      if (na.isConst())
+      {
+        Trace("lambda-const") << "*** Constant lambda : " << n;
+        Trace("lambda-const") << " since its array representation : " << na
+                              << " is constant." << std::endl;
+        return true;
+      }
+      else
+      {
+        Trace("lambda-const") << "Non-constant lambda : " << n
+                              << " since array is not constant." << std::endl;
+      }
+    }
+    else
+    {
+      Trace("lambda-const")
+          << "Non-constant lambda : " << n
+          << " since its varlist is not standard." << std::endl;
+      Trace("lambda-const") << "  standard : " << bvl << std::endl;
+      Trace("lambda-const") << "   current : " << n[0] << std::endl;
+    }
+  }
+  else
+  {
+    Trace("lambda-const") << "Non-constant lambda : " << n
+                          << " since it has no array representation."
+                          << std::endl;
+  }
+  return false;
+}
+
+Cardinality FunctionProperties::computeCardinality(TypeNode type)
+{
+  // Don't assert this; allow other theories to use this cardinality
```

Comment:
Is this comment needed?

Reply:
Yeah, it's saying the possible assertion doesn't hold, so I'd keep it.